### PR TITLE
Entity linking candidate generation based on char-n-gram and approximate nearest neighbors 

### PR DESCRIPTION
**Pull request by ibeltagy**
It works as follows: 
- map canonical names and aliases of entities in UMLS to sparse 1-hot encoding vectors using tfidf weighted char-3gram and char-4gram
- use `nmslib` to index all of UMLS
- given a query, use nmslib to search for `k` approximate nearest neighbors

Notes: 
- results are assuming gold NER
- prediction is very fast, but loading things in memory is not as fast
- the slowest part is vectorizing the KB (6 minutes), and the vectorized KB is more than 4GB (in case we prefer to load it from the disk). It will be great to find a way around this
- potential ways to speed things up: use char4gram only or don't use aliases. I will try both and see which one results it a smaller performance drop.

Data: 
- evaluation data: `s3://ai2-s2-scispacy/data/med_mentions.tar.gz`
- nmslib ann index: `s3://ai2-s2-scispacy/data/nmslib_index_char34gram_withaliases.bin`
- sklearn tfidf char34 gram vectorizer: `s3://ai2-s2-scispacy/data/tfidf_char34gram_vectorizer.joblib`
- umls release: `s3://ai2-s2-scispacy/data/umls_2017_aa_cat0129.json`

Running the following command
```python3 scripts/linking.py  --medmentions_path data/medmentions --umls_path data/umls_2017_aa_cat0129.json  --k 1,2,10,100,1000  --tfidf_vectorizer_path data/tfidfvec_char34gram_vectorizer.joblib --ann_index_path data/nmslib_index_char34gram_withaliases.bin```
Prints the following log:
```
Loading umls concepts from data/umls_2017_aa_cat0129.json
Number of umls concepts: 2783846
Collecting aliases ... 
Processed 1000000 or 2783846 concepts
Processed 2000000 or 2783846 concepts
Loading tfidf vectorizer from data/tfidfvec_char34gram_vectorizer.joblib
Vectorizing aliases ... 
Vectorizing aliases took 329.725663 seconds
Deleting 4289/6336776 aliases because their tfidf is empty
['MP' 'SL' 'SL' ... 'Ha' '3Y' '2Y']
Loading ann index from data/nmslib_index_char34gram_withaliases.bin
Reading MedMentions ... 
for k = 1
Generating candidates for 70306 mentions
Number of empty vectors: 1755
Finding neighbors took 25.768434 seconds
MedMentions entities not in UMLS: 756
MedMentions entities found in UMLS: 70306
K: 1
Gold concept in candidates: 54.92%
Gold concept not in candidates: 42.58%
Candidate generation failed: 2.50%
for k = 2
Generating candidates for 70306 mentions
Number of empty vectors: 1755
Finding neighbors took 23.073926 seconds
MedMentions entities not in UMLS: 756
MedMentions entities found in UMLS: 70306
K: 2
Gold concept in candidates: 63.45%
Gold concept not in candidates: 34.05%
Candidate generation failed: 2.50%
for k = 10
Generating candidates for 70306 mentions
Number of empty vectors: 1755
Finding neighbors took 25.654996 seconds
MedMentions entities not in UMLS: 756
MedMentions entities found in UMLS: 70306
K: 10
Gold concept in candidates: 73.53%
Gold concept not in candidates: 23.97%
Candidate generation failed: 2.50%
for k = 100
Generating candidates for 70306 mentions
Number of empty vectors: 1755
Finding neighbors took 31.947052 seconds
MedMentions entities not in UMLS: 756
MedMentions entities found in UMLS: 70306
K: 100
Gold concept in candidates: 81.68%
Gold concept not in candidates: 15.82%
Candidate generation failed: 2.50%
for k = 1000
Generating candidates for 70306 mentions
Number of empty vectors: 1755
Finding neighbors took 60.497052 seconds
MedMentions entities not in UMLS: 756
MedMentions entities found in UMLS: 70306
K: 1000
Gold concept in candidates: 85.40%
Gold concept not in candidates: 12.10%
Candidate generation failed: 2.50%
 ```

### CODE REVIEW

**Comment by ibeltagy** [comment]
CI failed. It could be related to this https://github.com/allenai/allennlp/pull/2647. How do you recommend fixing it ?

File: scripts/linking.py [hunk 1]
```diff
@@ -1,86 +1,241 @@
...
+
+    uml_concept_ids = np.array(uml_concept_ids)
+    uml_concept_aliases = np.array(uml_concept_aliases)
+    assert len(uml_concept_ids) == len(uml_concept_aliases)
+
+    if not os.path.isfile(tfidf_vectorizer_path):
+        print(f'No tfidf vectorizer on {tfidf_vectorizer_path}')
+        print(f'Fitting tfidf vectorizer on {len(uml_concept_aliases)} aliases')
+        tfidf_vectorizer = TfidfVectorizer(analyzer='char', ngram_range=(3, 4))
+        tfidf_vectorizer.fit(uml_concept_aliases)
```
Comment (DeNeutoy) [inline scripts/linking.py]: I believe you could combine these steps using `fit_transform`, I'm not sure how much of a speed boost you get - i'd imagine possibly substantial as here you are iterating twice over the collection.

File: scripts/linking.py [hunk 2]
```diff
@@ -1,86 +1,241 @@
...
+    start_time = datetime.datetime.now()
+    uml_concept_aliase_tfidfs = tfidf_vectorizer.transform(uml_concept_aliases)
+    end_time = datetime.datetime.now()
+    total_time = (end_time - start_time)
+    print(f'Vectorizing aliases took {total_time.total_seconds()} seconds')
+
+    # find empty (all zeros) tfidf vectors
+    empty_tfidfs_boolean_flags = np.array(uml_concept_aliase_tfidfs.sum(axis=1) != 0).reshape(-1,)
+    deleted_aliases = uml_concept_aliases[empty_tfidfs_boolean_flags == False]
+    number_of_none_empty_tfidfs = len(deleted_aliases)
```
Comment (DeNeutoy) [inline scripts/linking.py]: `number_of_non_empty_tfidfs`

File: scripts/linking.py [hunk 3]
```diff
@@ -1,86 +1,241 @@
...
+    total_number_of_tfidfs = uml_concept_aliase_tfidfs.shape[0]
+    print(f'Deleting {number_of_none_empty_tfidfs}/{total_number_of_tfidfs} aliases because their tfidf is empty')
+
+    # remove empty tfidf vectors, otherwise nmslib will crashd
+    uml_concept_ids = uml_concept_ids[empty_tfidfs_boolean_flags]
+    uml_concept_aliases = uml_concept_aliases[empty_tfidfs_boolean_flags]
+    uml_concept_aliase_tfidfs = uml_concept_aliase_tfidfs[empty_tfidfs_boolean_flags]
```
Comment (DeNeutoy) [inline scripts/linking.py]: `uml_concept_alias_tfidfs`

File: scripts/linking.py [hunk 4]
```diff
@@ -1,86 +1,241 @@
...
+            print(f'Processed {i} or {len(umls_concept_list)} concepts')
+
+    uml_concept_ids = np.array(uml_concept_ids)
+    uml_concept_aliases = np.array(uml_concept_aliases)
+    assert len(uml_concept_ids) == len(uml_concept_aliases)
+
+    if not os.path.isfile(tfidf_vectorizer_path):
+        print(f'No tfidf vectorizer on {tfidf_vectorizer_path}')
+        print(f'Fitting tfidf vectorizer on {len(uml_concept_aliases)} aliases')
+        tfidf_vectorizer = TfidfVectorizer(analyzer='char', ngram_range=(3, 4))
```
Comment (DeNeutoy) [inline scripts/linking.py]: Does the `char` vs `char_wb` distinction make a difference at all? My intuition would default to `char_wb`, which doesn't count ngrams across words.

Comment (DeNeutoy) [inline scripts/linking.py]: Have you played around with the `min_df` parameter? This might substantially reduce the dimensionality of the `CountVectorizer`. Also, have you examined performance using the `HashingVectorizer`? This might partially address the memory issues?

File: scripts/linking.py [hunk 5]
```diff
@@ -1,86 +1,241 @@
...
+    num_threads = 60
+    efS = 1000
+    index_params = {'M': M, 'indexThreadQty': num_threads, 'efConstruction': efC, 'post' : 0}
+
+    if not os.path.isfile(ann_index_path):
+        print(f'No ann index on {ann_index_path}')
+        print(f'Fitting ann index on {len(uml_concept_aliases)} aliases (takes 2 hours)')
+
+        start_time = datetime.datetime.now()
+        ann_index = nmslib.init(method='hnsw', space='cosinesimil_sparse', data_type=nmslib.DataType.SPARSE_VECTOR)
```
Comment (DeNeutoy) [inline scripts/linking.py]: Could you add some comments discussing the performance tradeoffs you've researched here? It would be good to know for future users why these settings are 1) as they are and 2) why the hyperparams are important.

File: scripts/linking.py [hunk 6]
```diff
@@ -1,86 +1,241 @@
...
 from scispacy import data_util
 
-def normalize(text):
-    return " ".join(text.lower().split(" "))
-
-def linking(entity: data_util.MedMentionEntity, umls_concept_dict_by_name):
+def load_umls_kb(umls_path: str):
     """
-    Links the entity mention and returns umls concept
+    Returns a list of concept object
```
Comment (DeNeutoy) [inline scripts/linking.py]: Please could you add return types, and make explicit what you are loading.

File: scripts/linking.py [hunk 7]
```diff
@@ -1,86 +1,241 @@
...
+    # neighbors need to be convected to an np.array of objects instead of ndarray of dimensions len(vectors)xk
+    # Solution: add a row to `neighbors` with any length other than k. This way, calling np.array(neighbors)
+    # returns an np.array of objects
+    neighbors.append([])
+    # interleave `neighbors` and Nones in `extended_neighbors`
+    extended_neighbors[empty_vectors_boolean_flags] = np.array(neighbors)[:-1]
+
+    return extended_neighbors
+
+def generate_candidates(mention_texts, k, tfidf_vectorizer, ann_index, ann_concept_id_list):
```
Comment (DeNeutoy) [inline scripts/linking.py]: types + return types.

File: scripts/linking.py [hunk 8]
```diff
@@ -1,86 +1,241 @@
...
+        k: number of ann neighbors
+        tfidf_vectorizer: text to vector
+        ann_index: approximate nearest neighbor index
+        ann_concept_id_list: a list of concept ids that the ann_index is referencing
     """
-    Returns two indices, one by entity id, and one by concept name.
+    print(f'Generating candidates for {len(mention_texts)} mentions')
+    tfidfs = tfidf_vectorizer.transform(mention_texts)
+    start_time = datetime.datetime.now()
+    neighbors = nmslis_knn_with_zero_vectors(tfidfs, k, ann_index)
```
Comment (DeNeutoy) [inline scripts/linking.py]: I don't understand how this works (the part where we do knn with zero vectors). Could you explain it a bit?

File: scripts/linking.py (hunk 4 above)
Comment (DeNeutoy) [reply scripts/linking.py]: Sorry, I mean the `max_df` parameter. Actually i'm pretty sure this is very important for the resulting size of the vectorizer, as it is dominated by vocabulary size, which can be significantly reduced by removing from consideration the long tail of ngrams.

File: scripts/linking.py (hunk 1 above)
Comment (ibeltagy) [reply scripts/linking.py]: True, but this is only relevant at the time of fitting the `TfidfVectorizer` (will save less than 5 minutes), but won't speed up prediction or loading to memory.

File: scripts/linking.py (hunk 4 above)
Comment (ibeltagy) [reply scripts/linking.py]: ok, will try `char_wb`

Comment (ibeltagy) [reply scripts/linking.py]: These are also worth checking. My intuition is that, 
- `min_df = 2` will reduce dimensionality significantly but will also hurt the performance a lot
- `max_df = 0.9` won't reduce the dimensionality that much
- `HashingVectorizer` is a good suggestion. I don't know how important the tfidf weighting is

**Comment by ibeltagy** [comment]
Addressed the comments above. You can probably use it now while I am experimenting with `char_wb`, `min_df`, `max_df` and `HashingVectorizer`.

**Comment by ibeltagy** [comment]
Ran the experiments and here are the results: 
- `char_wb`: better than `char`. 
- `min_df=10`: reduces vocabulary size and improves ann results a bit
- `max_df`: has no effect. Even with a low `max_df=0.2`, it doesn't affect vocabulary size
- Used char3gram only (instead of 3 and 4). This reduced vocabulary size from 500K to 50K, and reduced size of the tfidf matrix from 5GB to 3GB 
- with a tfidf mastix < 4GB, it is possible to save it on the disk and load it instead of reproducing it every time. 
- `HashingVectorizer`: results are usually 2.5% worse than TfidfVectorizer (performance drops even more with vocabulary size < 50K). More importantly, it doesn't reduce memory because the bottleneck is number of non-zero values in the array not the vocabulary size.

So, best configuration so far is: char_wb, char3gram, TfidfVectorizer, min_df=10. Loading everything in memory takes less than a minute, but requires many GBs of ram. 

Files: 
```s3://ai2-s2-scispacy/data/tfidfvec_char3gram_wbchar_vectorizer_mindf10.joblib.npy
s3://ai2-s2-scispacy/data/tfidfvec_char3gram_wbchar_vectorizer_mindf10.joblib
s3://ai2-s2-scispacy/data/nmslib_index_char3gram_withaliases_wbchar_mindf10.bin```

Then run 
```python3 scripts/linking.py --medmentions_path data/medmentions --umls_path data/umls_2017_aa_cat0129.json --k 1,2,10,100,1000 --tfidf_vectorizer_path data/tfidfvec_char3gram_wbchar_vectorizer_mindf10.joblib --ann_index_path data/nmslib_index_char3gram_withaliases_wbchar_mindf10.bin```

Prints the following log: 
```
Collecting aliases ...
Processed 1000000 or 2783846 concepts
Processed 2000000 or 2783846 concepts
No tfidf vectorizer on data/tfidfvec_char3gram_wbchar_vectorizer_mindf10.joblib
Fitting tfidf vectorizer on 6336776 aliases
Saving tfidf vectorizer to data/tfidfvec_char3gram_wbchar_vectorizer_mindf10.joblib
Saving tfidf vectors to data/tfidfvec_char3gram_wbchar_vectorizer_mindf10.joblib.npy
Loading tfidf vectorizer from data/tfidfvec_char3gram_wbchar_vectorizer_mindf10.joblib
Tfidf vocab size: 54337
Loading tfidf vectors from data/tfidfvec_char3gram_wbchar_vectorizer_mindf10.joblib.npy
Loading vectorizer and vectors took 24.519187 seconds
Deleting 7/6336776 aliases because their tfidf is empty
['F₃T' 't&a' 'g%' 'g%' 'g%' "K'" '/{VF}']
No ann index on data/nmslib_index_char3gram_withaliases_wbchar_mindf10.bin
Fitting ann index on 6336769 aliases (takes 2 hours)
 0%  10  20  30  40  50  60  70  80  90  100%
\|----\|----\|----\|----\|----\|----\|----\|----\|----\|----\|
***************************************************
Fitting ann index took 3282.471282 seconds
Loading ann index from data/nmslib_index_char3gram_withaliases_wbchar_mindf10.bin
Reading MedMentions ...
for k = 1
Generating candidates for 70306 mentions
Number of empty vectors: 32
Finding neighbors took 15.785411 seconds
MedMentions entities not in UMLS: 756
MedMentions entities found in UMLS: 70306
K: 1
Gold concept in candidates: 56.28%
Gold concept not in candidates: 43.68%
Candidate generation failed: 0.05%
for k = 2
Generating candidates for 70306 mentions
Number of empty vectors: 32
Finding neighbors took 15.46088 seconds
MedMentions entities not in UMLS: 756
MedMentions entities found in UMLS: 70306
K: 2
Gold concept in candidates: 64.31%
Gold concept not in candidates: 35.65%
Candidate generation failed: 0.05%
for k = 10
Generating candidates for 70306 mentions
Number of empty vectors: 32
Finding neighbors took 15.230339 seconds
MedMentions entities not in UMLS: 756
MedMentions entities found in UMLS: 70306
K: 10
Gold concept in candidates: 74.64%
Gold concept not in candidates: 25.31%
Candidate generation failed: 0.05%
for k = 100
Generating candidates for 70306 mentions
Number of empty vectors: 32
Finding neighbors took 16.611278 seconds
MedMentions entities not in UMLS: 756
MedMentions entities found in UMLS: 70306
K: 100
Gold concept in candidates: 83.28%
Gold concept not in candidates: 16.68%
Candidate generation failed: 0.05%
for k = 1000
Generating candidates for 70306 mentions
Number of empty vectors: 32
Finding neighbors took 33.373053 seconds
MedMentions entities not in UMLS: 756
MedMentions entities found in UMLS: 70306
K: 1000
Gold concept in candidates: 86.89%
Gold concept not in candidates: 13.06%
Candidate generation failed: 0.05%
```

**Comment by DeNeutoy** [comment]
This is great - 2% overall improvement and way smaller. I've also just noticed that the default `dtype` for the matrix returned by `TfidfVectorizer` is `numpy.float64`. It's pretty likely that doing Approximate Nearest Neighbours with `numpy.float16` vectors makes almost no difference, meaning that the tf-idf matrix might be < 1GB. Given that we can drop the dimensionality of the vocab from 500k -> 50k without any drop in performance, perhaps we should try using the `max_features=10k` or something, but use both 3 and 4 grams to compute the vocab. Combined with the float16 option, this might get us a tf-idf matrix < 100mb.

**Comment by ibeltagy** [comment]
Good suggestions
- `numpy.float16` is not supported by `scipy.sparse`. Changing `dtype` to `numpy.float32` reduced the matrix size from 3GB to 1.8GB. Will upload the new index after it is done building. 
- setting `max_features` reduced vocab size but doesn't significantly reduce number of nonzero values in the matrix because it drops the rare features. Setting it to `max_features=10000` reduces matrix size to 1.7GB
- `max_df` is good at reducing number of nonzero values (removes the most common features), but it quickly reduces performance. `max_df=100000`  reduces tfidf matrix size by half but reduces accuracy by 15%

**Comment by ibeltagy** [comment]
Uploaded 
```
s3://ai2-s2-scispacy/data/nmslib_index_char3gram_withaliases_wbchar_mindf10_float32.bin
s3://ai2-s2-scispacy/data/tfidfvec_char3gram_wbchar_vectorizer_mindf10_float32.joblib
s3://ai2-s2-scispacy/data/tfidfvec_char3gram_wbchar_vectorizer_mindf10_float32.joblib.npy
```

**Comment by ibeltagy** [comment]
Uploaded 
```
s3://ai2-s2-scispacy/data/data/nmslib_index.bin
s3://ai2-s2-scispacy/data/tfidfvec.joblib
s3://ai2-s2-scispacy/data/tfidfvec.joblib.npy
```
Performance is almost the same as float64
```
K = 1: 55.60%
K = 2: 64.01%
K = 10: 74.81%
K = 100: 83.31%
K = 1000: 86.96%
```
